api/src/main/java/org/openmrs/module/xdsbrepository/tasks/DiscreteDataProcessorTask.java

```diff
@@ -0,0 +1,119 @@
+package org.openmrs.module.xdsbrepository.tasks;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.hibernate.ObjectNotFoundException;
+import org.openmrs.EncounterRole;
+import org.openmrs.EncounterType;
+import org.openmrs.Patient;
+import org.openmrs.Provider;
+import org.openmrs.api.EncounterService;
+import org.openmrs.api.ProviderService;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.shr.contenthandler.api.Content;
+import org.openmrs.module.shr.contenthandler.api.ContentHandler;
+import org.openmrs.module.shr.contenthandler.api.ContentHandlerService;
+import org.openmrs.module.xdsbrepository.Utils;
+import org.openmrs.module.xdsbrepository.XDSbService;
+import org.openmrs.module.xdsbrepository.XDSbServiceConstants;
+import org.openmrs.module.xdsbrepository.exceptions.HydrateRoleProviderMapException;
+import org.openmrs.module.xdsbrepository.model.QueueItem;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class DiscreteDataProcessorTask extends Thread {
+
+    private Log log = LogFactory.getLog(DiscreteDataProcessorTask.class);
+
+    private int pollPeriod;
+
+    public DiscreteDataProcessorTask() {
+        pollPeriod = Integer.parseInt(Context.getAdministrationService().getGlobalProperty(
+                XDSbServiceConstants.XDS_REPOSITORY_DISCRETE_HANDLER_ASYNC_POLL_PERIOD, "200"));
+    }
+
+    @Override
+    public void run() {
+        QueueItem currentQueueItem;
+        XDSbService service = Context.getService(XDSbService.class);
+        Utils.startSession();
+
+        while (true) {
+            currentQueueItem = service.dequeueNextDiscreteDataForProcessing();
+            if (currentQueueItem != null) {
+                try {
+                    processQueueItem(currentQueueItem);
+                    service.completeQueueItem(currentQueueItem, true);
+                } catch (Exception e) {
+                    log.error("Error processing discrete data asynchronously for queue item "
+                            + currentQueueItem.getId() + " for documentUniqueId " + currentQueueItem.getDocUniqueId(), e);
+                    if (currentQueueItem != null) {
+                        service.completeQueueItem(currentQueueItem, false);
+                    }
```

Comment:
It looks like runtime exceptions won't be catered for.

Maybe instead of manually trying to manage the thread pooling it would be useful to use something like java's concurrent package; in particular there's the [ExecutorService](http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html) that can be used. But it's up to you.